### PR TITLE
279 - context menu entry "back to the star" (qt and gtk)

### PR DESCRIPTION
**Pull request by lrm25**

### CODE REVIEW

**Comment by lrm25** [comment]
Still needs the Windows commit, I don't have Visual Studio yet.

File: src/celestia/qt/qtselectionpopup.cpp [hunk 1]
```diff
@@ -180,20 +180,24 @@ SelectionPopup::SelectionPopup(const Selection& sel,
...
-    if (selection.body() != nullptr)
-    {
         // Alternate surface submenu
         QMenu* surfacesMenu = createAlternateSurfacesMenu();
         if (surfacesMenu != nullptr)
             addMenu(surfacesMenu);
 
+        if (selection.body()->hasPrimary())
+        {
+            QAction *action = new QAction(_("Primary"), this);
```
Comment (375gnu) [inline src/celestia/qt/qtselectionpopup.cpp]: I'd prefer a more descriptive title like "Select primary body"

File: src/celestia/gtk/menu-context.cpp [hunk 2]
```diff
@@ -66,6 +67,10 @@ void menuContext(float, float, Selection sel)
             AppendMenu(popup, NULL, "S_ync Orbit", gtk_action_group_get_action(app->agMain, "SyncSelection"));
             /* Add info eventually:
              * AppendMenu(popup, NULL, "_Info", 0); */
+            if (sel.body()->hasPrimary())
+            {
+                AppendMenu(popup, GTK_SIGNAL_FUNC(handleContextPrimary), "Primary", NULL);
```
Comment (375gnu) [inline src/celestia/gtk/menu-context.cpp]: The same as for Qt

File: src/celestia/qt/qtselectionpopup.h [hunk 3]
```diff
@@ -59,6 +60,10 @@ Q_OBJECT
     void selectionInfoRequested(Selection& sel);
 
  private:
+    bool hasPrimary(Body *body);
```
Comment (375gnu) [inline src/celestia/qt/qtselectionpopup.h]: it seems these 4 prototypes shouldn't be here

File: src/celengine/body.h [hunk 4]
```diff
@@ -361,6 +364,10 @@ class Body : public CatEntry
     void setName(const std::string& name);
     void recomputeCullingRadius();
 
+    bool hasPrimaryStar() const;
+    bool hasPrimaryBody(int classification) const;
+    bool hasPrimaryAroundBarycenter() const;
```
Comment (375gnu) [inline src/celengine/body.h]: just hasBarycenter()

File: src/celengine/body.h [hunk 5]
```diff
@@ -240,6 +240,9 @@ class Body : public CatEntry
     PlanetarySystem* getSatellites() const;
     void setSatellites(PlanetarySystem*);
 
+    bool hasPrimary() const;
+    CatEntry* getPrimary() const;
```
Comment (375gnu) [inline src/celengine/body.h]: I'm not sure that CatEntry* is a correct type here, I suppose Selection should be used here

File: src/celengine/body.cpp [hunk 6]
```diff
@@ -492,6 +492,90 @@ void Body::setSatellites(PlanetarySystem* ssys)
...
+        case Body::Comet:
+            return hasPrimaryStar();
+        case Body::Moon:
+        case Body::MinorMoon:
+            return hasPrimaryBody(Body::Planet | Body::DwarfPlanet) ||
+                   hasPrimaryAroundBarycenter();
+        case Body::Spacecraft:
+            return hasPrimaryBody(Body::Planet | Body::DwarfPlanet | Body::Asteroid |
+                                  Body::Comet | Body::Moon | Body::MinorMoon) ||
+                   hasPrimaryStar();
```
[the remark this answers is not in the document]
Comment (lrm25) [reply src/celengine/body.cpp]: How about "SpacecraftPrimaryBody"?

**Comment by lrm25** [comment]
> I have a general concern whether your new methods should be part of the `Body` class. Backend should supply minimal required data to a frontend, and then it's a frontend's job to process it. As we don't have yet any basic frontend class, then you can either move your code to celestia/celestiacore.cpp or add a new celestia/helper.cpp.

I'm thinking helper.cpp since the functions don't really seem to fit in celestiacore.cpp.

File: src/celengine/body.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (375gnu) [reply src/celengine/body.cpp]: > How about "SpacecraftPrimaryBody"?

Yes, this sound much better.

**Comment by lrm25** [comment]
For the helper module, would you prefer static functions or a class?

**Comment by 375gnu** [comment]
I'd prefer functions, the simpler architecture you help the better :)

**Comment by lrm25** [comment]
Let me fix the commits.

**Pull request closed without merging**